tools/parity/corpus/parameters.txt
14 %2BZoon%2B%28Leiden%29%2B%28uitvoerder%29
35 %2BVan%2Bder%2BLubbe%2B%28ontwerper%29
2 %20Cloppenburg%20%28Den%20Haag%29
10 %20Co.%20%28Amsterdam/Den%20Haag%29
357 %20N.%20Mutsaerts%20%28Tilburg%29
154 moreLikeThis.filterkey
3 sortBy%3Arandom
52 gewichten
47 waterketel
131 facet.limit
10 wvstest
324 scrollId
2 q.facet.sortReverse
1 %2BCo%2C%2BAmsterdam%2BFaddegon%22
82 A%29%20Ligtenberg
296 %2BZoon%2B%28Zeelst/Meerveldhoven%29
6 %20Vosseler%20Z%C3%A4hlerfabrik%20%28Schwenningen%29
95 %20Van%20der%20Lubbe
25 %20Zn
156 heiligenbeeld.
3 %2BCo.%2B%28Rijssen%29
526 %20Co%20%28Leverkusen%29
323 orderBy
1 %20Salzer%20Maschinenfabrik%20%28Ingolstadt%29
1 s
4 B%29%20%28Haarlem%29
84 %2BDagmar%2BDuquesne
6 %20Co.%20LTD.%20%28Bradford%29
133 %20Litermaat
3 %2BBinger%2BEmrik%22
1 %2BCompany
1 %2BFlentje%2B%28M%C3%BChlhausen%29%2B%28machinefabriek%29
3 %20Mutsaerts%20%28Tilburg%29
2 kioskrefresh
1 lt%22
249 %20Co
37 visklem
44 Snijmachine
46 wasuitrusting.
3 %2BZonen%2B%28Eindhoven%29%2B%28opdrachtgever%29
2 %20Sons%20Weaving%20Ltd.
2 echo
164 %2BDagmar%2BDuquesne%2B%28toegeschreven%29
82 %2BCo%2B%28Paisley%29
109 %20Schep
5 B
6 %20Campman%20%28Laren%29
4 facet.mincount
46478619 callback
126 %2Bco
14 %2BCo.%2B%28Reutlingen%29%2B%28machinefabriek%29
158 %20Leembruggen%20%28Leiden%29
1432 %2BCo.%2B%28Helmond%29
134 %2BCo.%2BSchlafhorst%2B%28M%C3%B6nchengladbach%29
1 co
211 %20Co.%20%28USA%29
20 %2BZn.%2B%28Geldrop%29
50 %20onderscheidingstekens
2 %2BP.%2BCoats
1 %2BCanvas%2BWorks%2B%28Londen%29
4 %2BA
843 %20ZONEN%20ROERMOND%E2%80%99.
3 facet_field
1 tib_collectionType_facet:Beeldmateriaal
169658997 hqf[]
12 %20Flentje%20%28M%C3%BChlhausen%29
5 %20Co.%20%28Rockford
3 Vluchtelingenwerk%20Brabant%20Centraal
19 %2BCo.%2B%28Almelo%29
1002 %2Bweegschaal
7 %2BZn%2B%28Oldenzaal%29
7 %2BCo.%2B%28Gent%29
652 %20Co.%20Amst.%E2%80%99
8 %20Zn.%20Enschede
3 %20Tilburg
80 %2BCie.%2B%28Tourcoing%29
84 %20Co.%20Schlafhorst%20%28M%C3%B6nchengladbach%29
4 %2BCo.%2B%28Leicester%29%2B%28machinefabrikant%29
22 %20Jansen%20%28Rotterdam%29
3 %2BCloppenburg
6 %2BN.%2BMutsaerts
5 6
1 %20T.%20Avery%20%28Birmingham%29%20%28machinefabrikant%29
139 %2BLeimer%2B%28Augsburg%29
138 %2BCo.%2B%28Einsingen%2Bbei%2BUlm%29
30 %2BCo.%2B%28Leiden%29%2B%28uitvoerder%29
13 %2BCo.%2BSchlafhorst%2B%28M%C3%B6nchengladbach%29%2B%28machinefabriek%29
14 %2BMeijers%2BBV%2BMeubelindustrie
1 %2BCarlier%2B%28Nuenen/Eeneind%29
2 K
1 %20Co.%20%28W%C3%A4denswil%29
5 %20Cie%20%28Krefeld%29
1 %20Co%20%28producent%29
1 %20Erfdijk%20%3A%20periodiek%20Heemkundekring%20Land%20van%20Ravenstein
64494354 mlt.filterkey
8116 qf[3]
213 %2BCo.%2B%28Reutlingen%29
4 15
3 %2BRaijmakers
2 %2BDams%2BTextielfabrieken%2B%28Tilburg%29
2 mizk_alle
82 %20Zoon%20%28Geldrop%29
2 %20P.%20Coats%20Ketting
1 Company%20%28Amsterdam%29
1 scroll_id
8 %20Co.%20%28Horgen%29
43 %2BZonen%2B%28Enschede%29
271 Schaal
4 %2BLauterbac%2BK%C3%B6rner%22
1 %2BGibbs%2BSewing%2BMachine%2BCo.%2B%28London%29%2B%28machinefabriek%29
7 %20Zonen%20%28Sleen%29
7 v1.m
46389410 diw-version
1 +lauterbac%22
421 %2BZonen%2B%28Waalre%29
3 qf%255B%255D
15 %2BCo.%2B%28W%C3%A4denswil%29
2 38
1 %2BZn.%2B%28Goirle%29
77 %20L.%20Burger%20%28
6 1f
2 diwView
3 qf[
2 W%201999
1123 %2BZonen%2B%28Tilburg%29
226 %2BHoldsworth%2B%28Enschede%29
10 %2BVan%2BPoppel%2B%28Tilburg%29%2B%28opdrachtgever%29
10 %2BCo.%2BN.V.
12 %2BBierens%2BRegenboog
1 rows500
39 cursor
343 %2BZn.%2B%28toegeschreven%29
20 %2BCo.%2B%28Sankt-Gallen%29%2B%28uitvoerder%29
5 %20Fils%20%28Frankrijk%29
922 %2BSchaal
153 %2BKruik
2 amp%3Bqf[0]
1034 %20Cie.%20Amsterdam%E2%80%99.
3 cb
67 brid
2 v1.format
1 n%20mutsaerts%20wollenstoffen%20tilburg
150 secret
54 %20Co.%20%28Utrecht%29
1024 %2BBinger
3 %20Bierens%20%28Tilburg%29
3 max_y
276 collection
1 %20Roelofs%20%28uitvoerder%29
1 %20Franca
187 T%20Textiles
102 %20Zonen%20%28toegeschreven%29
8 ampundefinedsortBy
4 %20Broel
46 %2BCo%2B%28Leverkusen%29
27 slijpgereedschap
5 %2BWeiss%2B%28Zeist%29
4 23
4 %20Holtus%20Terwindt%20%28Helmond%29
1 %20Co.%20N.V.%20%28Helmond%29
1 _query
2 D
346 %20Raijmakers%20%28Eindhoven%29
16 %2BZonen%2B%28Helmond%29
1 %2BZonen%27s%2BStoompassementfabriek%2B%28%27s-Hertogenbosch%29
5 %20Verster

CHANGELOG.md
# Changelog

## unreleased 

- history of changes: see https://github.com/delving/hub3/compare/v0.2.0...master

### Added

- ClI subcommand 'bulk' to index bulk-requests that are serialized to disk [[GH-88]](https://github.com/delving/hub3/pull/88)
- Config option to store records generated from METS files in dedicated index [[GH-83]](https://github.com/delving/hub3/pull/83)
- Resource package to ikuzo for uniform RDF handling [[GH-106]](https://github.com/delving/hub3/pull/106)

## Changed 
- 
- Allow for changes to uploaded EAD file before storing it [[GH-90]](https://github.com/delving/hub3/pull/90)
-  Refactored EAD search overview queries into separate filter and collapse queries [[GH-91]](https://github.com/delving/hub3/pull/91)
- Added support for NDE Dataset Register API [[GH-92]](https://github.com/delving/hub3/pull/92)
- Added posthooks support to orphan groups as well  [[GH-99]](https://github.com/delving/hub3/pull/99)

### Fixed
 
- Explicit initialisation of config for CLI subcommands instead of global on init [[GH-88]](https://github.com/delving/hub3/pull/88)
- Prevent duplicate files while processing METS files [[GH89]](https://github.com/delving/hub3/pull/89)
- Tree paging was using the old API key [[GH-93]](https://github.com/delving/hub3/pull/93)
- Delete non-EAD datasets would return incorrect error  [[GH-10]](https://github.com/delving/hub3/pull/102)

### Removed

- Unused embedded assets functionality in favour of `go:embed`  [[GH-99]](https://github.com/delving/hub3/pull/99)

### Security
- Resolved security issues in dependencies [[GH-95]](https://github.com/delving/hub3/pull/95)


## v0.2.0 (2021-03-10)

- history of changes: see https://github.com/delving/hub3/compare/v0.1.10...v0.2.0


### Added 

- Synchronize with remote archives and METS-files via OAI-PMH  [[GH-80]](https://github.com/delving/hub3/pull/80)
- Dedicated dataset service for managing datasets  [[GH-79]](https://github.com/delving/hub3/pull/79)

config/config.go
// Copyright 2017 Delving B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"encoding/json"
	"fmt"
	"log"
	"net/url"
	"os"
	"strings"

	"github.com/delving/hub3/ikuzo/logger"
	homedir "github.com/mitchellh/go-homedir"
	"github.com/rs/zerolog"
	"github.com/spf13/viper"
)

var (
	// Config is the general configuration object
	Config RawConfig

	// CfgFile is the path to the config file
	CfgFile string
)

// RawConfig holds all the configuration blocks.
// These are bound from cli, Environment variables or configuration files by
// Viper.
type RawConfig struct {
	HTTP          `json:"http"`
	ElasticSearch `json:"elasticsearch"`
	Logging       `json:"logging"`
	RDF           `json:"rdf"`
	OAIPMH        `json:"oaipmh"`
	WebResource   `json:"webresource"`
	ImageProxy    `json:"imageproxy"`
	LOD           `json:"lod"`
	RDFTag        `json:"rdftag"`
	PostHook      []PostHook `json:"postHook"`
	Cache         `json:"cache"`
	RDFTagMap     *RDFTagMap `json:"rdfTagMap"`
	EAD           `json:"ead"`
	DataSetTag    map[string]DataSets `json:"dataSetTag"`
	DatasetTagMap *RDFTagMap
	Logger        *zerolog.Logger `json:"logger"`
}

type DataSets struct {
	Specs []string
}

// PostHook contains the configuration for the JSON-LD posthook configuration
type PostHook struct {
	Name        string   `json:"name"`
	ExcludeSpec []string `json:"excludeSpec"`
	URL         string   `json:"url"`
	OrgID       string   `json:"orgID"`
	APIKey      string   `json:"apiKey"`
	UserName    string   `json:"userName"`
	Password    string   `json:"password"`
	CustomWait  int      `json:"customWait"`
}

// ElasticSearch holds all the configuration values
// It is bound by Viper.
type ElasticSearch struct {
	Urls                []string `json:"urls"`
	Enabled             bool     `json:"enabled"`
	IndexName           string   `json:"indexName"`
	DigitalObjectSuffix string   `json:"digitalObjectSuffix"`
	Proxy               bool     `json:"proxy"`
	Fragments           bool     `json:"fragments"`
	IndexV1             bool     `json:"indexV1"` // exclusive with v2 indexing
	EnableTrace         bool     `json:"enableTrace"`
	EnableInfo          bool     `json:"enableInfo"`
	SpecKey             string   `json:"specKey"`
	RevisionKey         string   `json:"revisionKey"`
	OrgIDKey            string   `json:"orgIDKey"`
	UserName            string   `json:"userName"`
	Password            string   `json:"password"`
	FacetSize           int      `json:"facetSize"`
	MinimumShouldMatch  string   `json:"minimumShouldMatch"`
	Workers             int      `json:"workers"`
	Shards              int      `json:"shards"`
	Replicas            int      `json:"replicas"`
	RequestTimeout      int      `json:"requestTimeout"`
	EnableSearchAfter   bool     `json:"enableSearchAfter"`
	TrackTotalHits      bool     `json:"trackTotalHits"`
	IndexTypes          []string `json:"indexTypes"`
	MaxTreeSize         int      `json:"maxTreeSize"`
	OrphanWait          int      `json:"orphanWait"`
	SearchFields        []string `json:"searchFields"`
	// OrphanTagsFilter enables filtering orphan deletion by meta.tags (mdr, narthex, ead, eadDesc).
	// Historically this was always enabled, but it causes orphans to persist when records
	// don't have these tags. Set to false (default) to delete orphans based only on spec+orgID.
	OrphanTagsFilter bool `json:"orphanTagsFilter"`
	// IndexingNotification configures webhook notifications for indexing events
	IndexingNotification IndexingNotificationConfig `json:"indexingNotification"`
}

// IndexingNotificationConfig holds configuration for webhook notifications
// sent during indexing operations (success, warnings, errors)
type IndexingNotificationConfig struct {
	Enabled  bool   `json:"enabled"`
	Endpoint string `json:"endpoint"`
	Timeout  int    `json:"timeout"` // timeout in seconds for webhook requests
	APIKey   string `json:"apiKey"`  // API key for Bearer token authentication
}

func (es *ElasticSearch) FragmentIndexName(orgID string) string {
	return fmt.Sprintf("%s_frag", es.GetIndexName(orgID))
}

func (es *ElasticSearch) GetSuggestIndexName(orgID string) string {
	return fmt.Sprintf("%s_suggest", es.GetIndexName(orgID))
}

// HasAuthentication returns if ElasticSearch has authentication enabled.
func (es *ElasticSearch) HasAuthentication() bool {
	return len(es.UserName) > 0 && len(es.Password) > 0
}

// GetIndexName returns the lowercased indexname.
// This inforced correct behavior when creating an index in ElasticSearch.
func (es *ElasticSearch) GetIndexName(orgID string) string {
	return strings.ToLower(orgID) + "v2"
}

func (es *ElasticSearch) GetV1IndexName(orgID string) string {
	return strings.ToLower(orgID) + "v1"
}

func (es *ElasticSearch) GetDigitalObjectIndexName(orgID string) string {
	if es.DigitalObjectSuffix == "" {
		return es.GetIndexName(orgID)
	}

	return strings.ToLower(orgID) + "v2-" + strings.ToLower(es.DigitalObjectSuffix)
}

// Logging holds all the logging and path configuration
type Logging struct {
	DevMode       bool   `json:"devmode"`
	SentryDSN     string `json:"sentrydsn"`
	Level         string `json:"level"`
	WithCaller    bool   `json:"withCaller"`
	ConsoleLogger bool   `json:"consoleLogger"`
}

// HTTP holds all the configuration for the http server subcommand
type HTTP struct {
	Port      int    `json:"port" mapstructure:"port"`
	StaticDir string `json:"staticDir"` // the relative path to the static directory to serve documentation.
	ProxyTLS  bool   `json:"proxyTLS"`
}

// RDF holds all the configuration for SPARQL queries and RDF conversions
type RDF struct {
	SparqlEnabled    bool   `json:"sparqlEnabled"`    // Enable the SPARQL proxy
	SparqlHost       string `json:"sparqlHost"`       // the base-url to the SPARQL endpoint including the scheme and the port
	SparqlPath       string `json:"sparqlPath"`       // the relative path of the endpoint. This can should contain the database name that is injected when the sparql endpoint is build
	SparqlUpdatePath string `json:"sparqlUpdatePath"` // the relative path of the update endpoint. This can should contain the database name that is injected when the sparql endpoint is build
	GraphStorePath   string `json:"dataPath"`         // the relative GraphStore path of the endpoint. This can should contain the database name that is injected when the sparql endpoint is build
	BaseURL          string `json:"baseUrl"`          // the RDF baseUrl used for minting new URIs (should not include scheme)
	BaseScheme       string `json:"baseScheme"`       // the scheme (http or https) used in the baseURL
	RDFStoreEnabled  bool   `json:"rdfStoreEnabled"`  // Store to Triple Store while saving RDF
	// the RDF entryPoints. Lookups are made on the fully qualified URIs. It is sometimes needed to support other baseUrls as well.
	// The entry-points need to be fully qualified, i.e. with their scheme.
	RoutedEntryPoints []string `json:"RoutedEntryPoints"`
	Tags              string   `json:"tags" mapstructure:"tags"`
	DefaultFormat     string   `json:"defaultFormat"`
	RDFStoreTags      []string `json:"rdfStoreTags"` // the tags that trigger storage in the triple-store
	UserName          string   `json:"userName"`     // username for SPARQL endpoint authentication
	Password          string   `json:"password"`     // password for SPARQL endpoint authentication
}

// HasAuthentication returns if RDF SPARQL endpoint has authentication enabled.
func (rdf *RDF) HasAuthentication() bool {
	return len(rdf.UserName) > 0 && len(rdf.Password) > 0
}

func (rdf *RDF) HasStoreTag(tags []string) bool {
	for _, tag := range tags {
		for _, storeTag := range rdf.RDFStoreTags {
			if strings.EqualFold(tag, storeTag) {
				return true
			}
		}
	}

	return false
}

// Cache is the configuration of the BigCache implementation
type Cache struct {
	Enabled           bool   `json:"enabled"`
	LifeWindowMinutes int    `json:"lifeWindowMinutes"`
	APIPrefix         string `json:"apiPrefix"`
	StripPrefix       bool   `json:"stripPrefix"`
	CacheDomain       string `json:"cacheDomain"`
	HardMaxCacheSize  int    `json:"hardMaxCacheSize"`
	MaxEntrySize      int    `json:"maxEntrySize"`
}

// OAIPMH holds all the configuration options for the OAI-PMH endpoint
type OAIPMH struct {
	// Make the oai-pmh endpoint available
	Enabled bool `json:"enabled"`
	// AdminEmails has a list of the admin emails of this endpoint
	AdminEmails []string `json:"adminEmails"`
	// RepositoryName is the name of the OAI-PMH repossitory
	RepositoryName  string `json:"repositoryName"`
	HarvestDelay    int    `json:"harvestDelay"`
	EadHarvestURL   string `json:"eadHarvestURL"`
	MetsHarvestURL  string `json:"metsHarvestURL"`
	MetsDownloadFmt string `json:"metsDownloadFmt"`
}

// WebResource holds all the configuration options for the WebResource endpoint
type WebResource struct {
	Enabled                   bool   `json:"enabled"`          // Make the webresource endpoint available
	WebResourceDir            string `json:"sourceDir"`        // Target directory for the webresources
	CacheResourceDir          string `json:"cacheDir"`         // cache directory for the webresources
	enableSearch              bool   `json:"enableSearch"`     // enable searching for images in ElasticSearch
	MediaManagerHost          string `json:"mediaManagerHost"` // the domain to build the derivatives
	SmallDefault              string `json:"smallDefault"`
	MediumDefault             string `json:"mediumDefault"`
	LargeDefault              string `json:"largeDefault"`
	MaxSize                   int32  `json:"maxSize"`
	ResolveRemoteWebResources bool   `json:"resolveRemoteWebResources"`
}

// ImageProxy holds all the configuration for the ImageProxy functionality
type ImageProxy struct {
	Enabled          bool     `json:"enabled"`          // Make the imageproxy endpoint available
	CacheDir         string   `json:"cacheDir"`         // The path to the imageCache
	Referrer         []string `json:"referrer"`         // A list of allowed referrers. If empty allow all.
	Whitelist        []string `json:"whitelist"`        // A list of allowed remote hosts. If empty allow all.
	ScaleUp          bool     `json:"scaleUp"`          // Allow images to scale beyond their original dimensions.
	TimeOut          int      `json:"timeout"`          // timelimit for request served by this proxy. 0 is for no timeout
	Deepzoom         bool     `json:"deepzoom"`         // Enable deepzoom of remote images.
	ProxyPrefix      string   `json:"proxyPrefix"`      // The prefix where we mount the imageproxy. default: imageproxy. default: imageproxy.
	DefaultImagePath string   `json:"defaultImagePath"` // Default image in case the image cannot be found.
	VipsPath         string   `json:"vipsPath"`         // Custom path to vips binary (empty for system PATH)
	VipsCliPath      string   `json:"vipsCliPath"`      // Custom path to vips-cli binary (empty to disable)
	ForceIPv4        bool     `json:"forceIPv4"`        // Force IPv4 for outgoing requests (some hosts block IPv6)
}

// LOD holds all the configuration for the Linked Open Data (LOD) functionality
type LOD struct {
	Enabled           bool   `json:"enabled"`        // Make the lod endpoint available
	Resource          string `json:"resource"`       // the 303 redirect entry point. This is where the content negotiation happens
	HTML              string `json:"html"`           // the endpoint that renders the data as formatted HTML
	RDF               string `json:"rdf"`            // the endpoint that renders the RDF data in the requested RDF format. Currently, JSON-LD and N-triples are supported
	SingleEndpoint    string `json:"singleEndpoint"` // when this is set it overrides the other endpoints
	HTMLRedirectRegex string `json:"redirectregex"`  // the regular expression to convert the subject uri to the uri for the external Page view
	Store             string // default sparql otherwise fragments
}

// SiteMap holds all the configuration for the sitemap generation
type SiteMap struct {
	Enabled bool   `json:"enabled"`
	BaseDir string `json:"baseDir"`
	BaseURL string `json:"baseURL"`
	Gzip    bool   `json:"gzip"`
}

// EAD holds all the configuration for the EAD endpoint
type EAD struct {
	CacheDir         string   `json:"cacheDir"`
	SearchURL        string   `json:"searchURL"`
	GenreFormDefault string   `json:"genreFormDefault"`
	TreeFields       []string `json:"treeFields"`
	SearchFields     []string `json:"searchFields"`
	Genreforms       []string `json:"genreforms"`
}

func setDefaults() {
	// setting defaults
	viper.SetDefault("HTTP.port", 3001)
	viper.SetDefault("HTTP.staticDir", "public")
	viper.SetDefault("orgId", "hub3")

	// elastic
	viper.SetDefault("ElasticSearch.urls", []string{"http://localhost:9200"})
	viper.SetDefault("ElasticSearch.enabled", true)
	viper.SetDefault("ElasticSearch.IndexName", viper.GetString("OrgId"))
	viper.SetDefault("ElasticSearch.Proxy", true)
	viper.SetDefault("ElasticSearch.Fragments", true)
	viper.SetDefault("ElasticSearch.IndexV1", false)
	viper.SetDefault("ElasticSearch.EnableTrace", false)
	viper.SetDefault("ElasticSearch.EnableInfo", false)
	viper.SetDefault("ElasticSearch.SpecKey", "meta.spec")
	viper.SetDefault("ElasticSearch.RevisionKey", "meta.revision")
	viper.SetDefault("ElasticSearch.OrgIDKey", "meta.orgID")
	viper.SetDefault("ElasticSearch.FacetSize", 50)
	viper.SetDefault("ElasticSearch.MimimumShouldMatch", "2<70%")
	viper.SetDefault("ElasticSearch.workers", 1)
	viper.SetDefault("ElasticSearch.EnableSearchAfter", false)
	viper.SetDefault("ElasticSearch.Shards", 1)
	viper.SetDefault("ElasticSearch.Replicas", 0)
	viper.SetDefault("ElasticSearch.RequestTimeout", 15)
	viper.SetDefault("ElasticSearch.MaxTreeSize", 250)
	viper.SetDefault("ElasticSearch.OrphanWait", 30)
	viper.SetDefault("ElasticSearch.TrackTotalHits", true)
	viper.SetDefault("ElasticSearch.IndexTypes", []string{"v2"})
	viper.SetDefault("ElasticSearch.IndexingNotification.Enabled", false)
	viper.SetDefault("ElasticSearch.IndexingNotification.Endpoint", "")
	viper.SetDefault("ElasticSearch.IndexingNotification.Timeout", 10)
	viper.SetDefault("ElasticSearch.IndexingNotification.APIKey", "")

	// logging
	viper.SetDefault("Logging.DevMode", false)
	viper.SetDefault("Logging.level", "info")

	// cache
	viper.SetDefault("Cache.enabled", false)
	viper.SetDefault("Cache.lifeWindow", 10)
	viper.SetDefault("Cache.apiPrefix", "/api/http/cache")
	viper.SetDefault("Cache.stripPrefix", true)
	viper.SetDefault("Cache.cacheDomain", "")
	viper.SetDefault("Cache.HardMaxCacheSize", 512)
	viper.SetDefault("Cache.MaxEntrySize", 10240)

	// rdf with defaults for Apache Fuseki
	viper.SetDefault("RDF.SparqlEnabled", true)
	viper.SetDefault("RDF.SparqlHost", "http://localhost:3030")
	viper.SetDefault("RDF.SparqlPath", "/%s/sparql")
	viper.SetDefault("RDF.SparqlUpdatePath", "/%s/update")
	viper.SetDefault("RDF.GraphStorePath", "/%s/data")
	viper.SetDefault("RDF.BaseUrl", "http://data.hub3.org")
	viper.SetDefault("RDF.BaseScheme", "http")
	viper.SetDefault("RDF.RoutedEntryPoints", []string{"http://localhost:3000", "http://localhost:3001"})
	viper.SetDefault("RDF.RDFStoreEnabled", false)
	viper.SetDefault("RDF.DefaultFormat", "application/ld+json")
	viper.SetDefault("RDF.RDFStoreTags", []string{"narthex", "mdr"})

	// rdftags
	viper.SetDefault("RDFTag.Label", []string{
		"http://www.w3.org/2004/02/skos/core#prefLabel",
		"http://purl.org/dc/elements/1.1/title",
	})

	// posthook
	viper.SetDefault("PostHook.urls", []string{})
	viper.SetDefault("PostHook.ExcludeSpec", []string{})

	// oai-pmh
	viper.SetDefault("OAIPMH.enabled", true)
	viper.SetDefault("OAIPMH.RepositoryName", "hub3")
	viper.SetDefault("OAIPMH.AdminEmails", "info@delving.eu")
	viper.SetDefault("OAIPMH.HarvestDelay", 1)
	viper.SetDefault("OAIPMH.EadHarvestUrl", "")
	viper.SetDefault("OAIPMH.MetsHarvestUrl", "")

	// image proxy
	viper.SetDefault("ImageProxy.enabled", true)
	viper.SetDefault("ImageProxy.CacheDir", "webresource_cache/cache")
	viper.SetDefault("ImageProxy.referrer", []string{})
	viper.SetDefault("ImageProxy.whitelist", []string{})
	viper.SetDefault("ImageProxy.scaleUp", false)
	viper.SetDefault("ImageProxy.timeout", 0)
	viper.SetDefault("ImageProxy.deepzoom", true)
	viper.SetDefault("ImageProxy.proxyPrefix", "imageproxy")
	viper.SetDefault("ImageProxy.DefaultImagePath", "")
	viper.SetDefault("ImageProxy.VipsPath", "")
	viper.SetDefault("ImageProxy.VipsCliPath", "")

	// sitemap
	viper.SetDefault("Sitemap.enabled", false)

	// webresource
	viper.SetDefault("WebResource.enabled", true)
	viper.SetDefault("WebResource.WebResourceDir", "/tmp/webresource")
	viper.SetDefault("WebResource.CacheResourceDir", "/tmp/webresource_cache")
	viper.SetDefault("WebResource.enableSearch", false)
	viper.SetDefault("WebResource.resolveRemoteWebResources", false)
	viper.SetDefault("WebResource.mediaManagerHost", "https://media.delving.org")
	viper.SetDefault("WebResource.SmallDefault", "220")
	viper.SetDefault("WebResource.MediumDefault", "350")
	viper.SetDefault("WebResource.LargeDefault", "500")
	viper.SetDefault("WebResource.MaxSize", int32(2000))

	// lod
	viper.SetDefault("LOD.enabled", true)
	viper.SetDefault("LOD.html", "page")
	viper.SetDefault("LOD.rdf", "data")
	viper.SetDefault("LOD.singleEndpoint", "")
	viper.SetDefault("LOD.resource", "resource")
	viper.SetDefault("LOD.redirectregex", "")

	// sitemap
	viper.SetDefault("SiteMap.Enabled", false)

	// pid 
	viper.SetDefault("PID.enabled", false)

	// ead
	viper.SetDefault("EAD.CacheDir", "/tmp/ead")
}

func cleanConfig() {
	Config.RDF.BaseURL = strings.TrimSuffix(Config.RDF.BaseURL, "/")
	if !strings.HasPrefix(Config.RDF.BaseScheme, "http") {
		Config.Logger.Fatal().
			Str("baseURL", Config.RDF.BaseURL).
			Msg("RDF.BaseUrl config value '%s' should start with 'http' or 'https'.")
	}
}

func SetCfgFile(cfgFile string) {
	CfgFile = cfgFile
}

// InitConfig reads in config file and ENV variables if set.
func InitConfig() {
	// InitConfig() must be idempotent
	if Config.Logger != nil {
		return
	}

	if CfgFile != "" {
		// Use config file from the flag.
		viper.SetConfigFile(CfgFile)
	} else {
		// Find home directory.
		home, err := homedir.Dir()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}

		// Search config in home directory with name "hub3" (without extension).
		viper.AddConfigPath("/etc/default/")
		viper.AddConfigPath(home)
		viper.AddConfigPath(".")
		viper.SetConfigName("hub3")
	}

	viper.SetEnvPrefix("HUB3")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	viper.AutomaticEnv() // read in environment variables that match

	setDefaults()

	logCfg := logger.Config{
		LogLevel:            logger.ParseLogLevel(Config.Logging.Level),
		WithCaller:          Config.Logging.WithCaller,
		EnableConsoleLogger: Config.Logging.ConsoleLogger,
	}

	configLogger := logger.NewLogger(logCfg)
	Config.Logger = &configLogger.Logger

	// If a config file is found, read it in.
	err := viper.ReadInConfig()
	if err == nil {
		Config.Logger.Info().
			Str("configPath", viper.ConfigFileUsed()).
			Msg("starting up with config path (legacy)")
	} else {
		log.Printf("Unable to read config file %s", viper.ConfigFileUsed())
		switch v := err.(type) {
		case viper.ConfigParseError:
			log.Fatalf("config parse error: %#v", v.Error())
		default:
			log.Printf("config parse error: %#v", err)
		}
	}

	err = viper.Unmarshal(&Config)
	if err != nil {
		log.Fatalf("unable to decode into struct, %v", err)
	}

	Config.RDFTagMap = NewRDFTagMap(&Config)
	Config.DatasetTagMap = NewDataSetTagMap(&Config)

	cleanConfig()
}

// GetSparqlEndpoint builds the SPARQL endpoint from the RDF Config object.
// When the dbName is empty the OrgId from the configuration is used.
func (c *RawConfig) GetSparqlEndpoint(orgID, dbName string) string {
	if dbName == "" {
		dbName = orgID
	}

	u, err := url.Parse(c.RDF.SparqlHost)
	if err != nil {
		log.Fatal(err)
	}

	u.Path = fmt.Sprintf(c.RDF.SparqlPath, dbName+"-data")

	return u.String()
}

// GetSparqlUpdateEndpoint builds the SPARQL Update endpoint from the RDF Config object.
// When the dbName is empty the OrgId from the configuration is used.
func (c *RawConfig) GetSparqlUpdateEndpoint(orgID, dbName string) string {
	if dbName == "" {
		dbName = orgID
	}

	u, err := url.Parse(c.RDF.SparqlHost)
	if err != nil {
		log.Fatal(err)
	}

	u.Path = fmt.Sprintf(c.RDF.SparqlUpdatePath, dbName+"-data")

	return u.String()
}

// GetGraphStoreEndpoint builds the GraphStore endpoint from the RDF Config object.
// When the dbName is empty the OrgId from the configuration is used.
func (c *RawConfig) GetGraphStoreEndpoint(orgID, dbName string) string {
	if dbName == "" {
		dbName = orgID
	}

	u, err := url.Parse(c.RDF.SparqlHost)
	if err != nil {
		log.Fatal(err)
	}

	u.Path = fmt.Sprintf(c.RDF.GraphStorePath, dbName+"-data")

	log.Printf("GraphStore endpoint: %s", u)

	return u.String()
}

// Save saves the update version of the configuration file
// At the moment this is mostly used for persisting the namespaces
func (c *RawConfig) Save() error {
	return viper.SafeWriteConfig()
}

// BuildVersionInfo holds all the version information
type BuildVersionInfo struct {
	Version    string `json:"version"`
	Commit     string `json:"commit"`
	BuildAgent string `json:"buildAgent"`
	BuildDate  string `json:"buildDate"`
	BuildID    string `json:"buildID"`
}

// NewBuildVersionInfo creates a BuildVersionInfo struct
func NewBuildVersionInfo(version, commit, buildagent, builddate string) *BuildVersionInfo {
	if version == "" {
		version = "devBuild"
	}

	return &BuildVersionInfo{
		Version:    version,
		Commit:     commit,
		BuildAgent: buildagent,
		BuildDate:  builddate,
	}
}

// JSON returns a json version of the BuildVersionInfo
func (b *BuildVersionInfo) JSON(pretty bool) ([]byte, error) {
	if pretty {
		return json.MarshalIndent(b, "", "\t")
	}

	return json.Marshal(b)
}
